modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/filters/DynamicFilter.js

```diff
@@ -0,0 +1,281 @@
+		actionType === 'delete' ||
+		((!selectedData || !selectedData.from) && fromValue) ||
+		((!selectedData || !selectedData.to) && toValue) ||
+		(selectedData &&
+			selectedData.from &&
+			fromValue !== formatDateObject(selectedData.from)) ||
+		(selectedData &&
+			selectedData.to &&
+			toValue !== formatDateObject(selectedData.to))
+	) {
+		submitDisabled = false;
+	}
+
+	return (
+		<>
+			<ClayDropDown.Caption>
+				<div className="form-group">
+					<ClayForm.Group className="form-group-sm">
+						<label htmlFor={`from-${id}`}>
+							{Liferay.Language.get('from')}
+						</label>
+
+						<input
+							className="form-control"
+							id={`from-${id}`}
+							max={toValue || (max && formatDateObject(max))}
+							min={min && formatDateObject(min)}
+							onChange={(event) =>
+								setFromValue(event.target.value)
+							}
+							pattern="\d{4}-\d{2}-\d{2}"
+							placeholder={placeholder || 'yyyy-mm-dd'}
+							type="date"
+							value={fromValue || ''}
+						/>
+					</ClayForm.Group>
+
+					<ClayForm.Group className="form-group-sm mt-2">
+						<label htmlFor={`to-${id}`}>
+							{Liferay.Language.get('to')}
+						</label>
+
+						<input
+							className="form-control"
+							id={`to-${id}`}
+							max={max && formatDateObject(max)}
+							min={fromValue || (min && formatDateObject(min))}
+							onChange={(event) => setToValue(event.target.value)}
+							pattern="\d{4}-\d{2}-\d{2}"
+							placeholder={placeholder || 'yyyy-mm-dd'}
+							type="date"
+							value={toValue || ''}
+						/>
+					</ClayForm.Group>
+				</div>
+			</ClayDropDown.Caption>
+			<ClayDropDown.Divider />
+			<ClayDropDown.Caption>
+				<ClayButton
+					disabled={submitDisabled}
+					onClick={() => {
+						if (actionType === 'delete') {
+							setFilter({active: false, id});
+						}
+						else {
+							const newSelectedData = {
+								from: fromValue
+									? getDateFromDateString(fromValue)
+									: null,
+								to: toValue
+									? getDateFromDateString(toValue)
+									: null,
+							};
+
+							setFilter({
+								active: true,
+								id,
+								odataFilterString: getOdataString({
+									entityFieldType,
+									id,
+									selectedData: newSelectedData,
+								}),
+								selectedData: newSelectedData,
+								selectedItemsLabel: getSelectedItemsLabel({
+									selectedData: newSelectedData,
+								}),
+							});
+						}
+					}}
+					small
+				>
+					{actionType === 'add' && Liferay.Language.get('add-filter')}
+
+					{actionType === 'edit' &&
+						Liferay.Language.get('edit-filter')}
+
+					{actionType === 'delete' &&
+						Liferay.Language.get('delete-filter')}
+				</ClayButton>
+			</ClayDropDown.Caption>
+		</>
+	);
+    */
+};
+
+DynamicFilter.propTypes = {
+	id: PropTypes.string.isRequired,
+	listTypeDefinitionId: PropTypes.string,
+	multiple: PropTypes.bool,
+	preselectedValues: PropTypes.arrayOf(PropTypes.string),
+	selectedData: PropTypes.arrayOf(PropTypes.string),
```

Comment:
Whats the difference here between these two?

Reply:
`preselectedValues` comes from the server, whereas `selectedData` is part of the internal state and can also come as initial data from the filters UI.

Reply:
Where does `selectedData` come from? Is there a parent component that is setting that? Because this is a prop, so it isn't internal state to this component at least

Reply:
`preselectedValues` comes from [here](https://github.com/liferay/liferay-portal/blob/e28a22f64fda1bfb7fda2dc8a6ed4c0668481f62/modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java#L426)

The "root" component comes [from the server](https://github.com/liferay/liferay-portal/blob/e28a22f64fda1bfb7fda2dc8a6ed4c0668481f62/modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/java/com/liferay/frontend/data/set/views/web/internal/fragment/renderer/FDSViewFragmentRenderer.java#L207).

The parent of the `DynamicFilter` component is [the Filter component](https://github.com/liferay/liferay-portal/blob/e28a22f64fda1bfb7fda2dc8a6ed4c0668481f62/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/filters/Filter.js#L58).

And the `selectedData` comes from some upper level between `FrontendDataSet` and `Filter`... Specifically, from these files:

![image](https://github.com/liferay-frontend/liferay-portal/assets/3273630/b4c32351-3279-40b8-8e8d-5d8d411c6030)

So, `selectedData` is a contract between the modular filter component (`DynamicFilter`) and the `FrontendDataSet`. It is created by the filter when the user populates the UI, passed to the `FrontendDataSet` using `setFilter` and `FrontendDataSet` becomes the owner and manager of that data, using it to filter the table, to show the active filters at the top of the table, and to render the filter dialog whenever it's needed again, with the currently active options.

Reply:
Oh I see. Yeah that makes more sense. I think `setFilter` will just always initially throw me off